Pendulum_with_elastic_balls.rst

```diff
@@ -0,0 +1,540 @@
+    FB = []
+    for i in range(n):
+        for j in range(i+1, n):
+            aa = Dmc[j].pos_from(Dmc[i])
+            bb = aa.magnitude()
+            aa = aa.normalize()
+            forceij = (Dmc[j],  k * (2*r - bb) * aa * sm.Heaviside(2.*r - bb))
+            FB.append(forceij)
+            forceji = (Dmc[i], -k * (2*r - bb) * aa * sm.Heaviside(2.*r - bb))
+            FB.append(forceji)
+    
+    FL = FG + FB       # list of forces
+    
+    # kinematic equations
+    kd = []
+    for i in range(n):
+        for uv in A[i]:
+            kd.append(me.dot(rot[i] - rot1[i], uv))
+        
+    # Kanes's equations
+    q1 = q
+    u1 = u
+    
+    KM = me.KanesMethod(N, q_ind=q1, u_ind=u1, kd_eqs=kd)
+    (fr, frstar) = KM.kanes_equations(BODY, FL)
+    
+    MM = KM.mass_matrix_full
+    print('MM DS', me.find_dynamicsymbols(MM))
+    print('MM free symbols', MM.free_symbols)
+    print('MM contains {} operations'.format(sum([MM[i, j].count_ops(visual=False) 
+            for i in range(MM.shape[0]) for j in range(MM.shape[1])])), '\n')
+    
+    force = KM.forcing_full
+    print('force DS', me.find_dynamicsymbols(force))
+    print('force free symbols', force.free_symbols)
+    print('force contains {} operations'.format(sum([force[i].count_ops(visual=False) 
+            for i in range(force.shape[0])])), '\n')
+    
+    # set up the energy equations. Absent ary friction the total energie should be cnstant
+    pot_energie = sum([m*g*me.dot(Dmc[i].pos_from(P[0]), N.y) for i in range(n)]) + sum([m1*g*me.dot(punkt[i]
+                            .pos_from(P[0]), N.y) for i in range(n)]) 
+    kin_energie = sum([BODY[i].kinetic_energy(N) for i in range(2*n)]) 
+    spring_energie = sm.S(0.)
+    for i in range(n):
+        for j in range(i+1, n):
+            aa = Dmc[j].pos_from(Dmc[i])
+            bb = aa.magnitude()
+            aa = aa.normalize()
+            spring_energie  += 0.5 * k * (2*r - bb)**2 * sm.Heaviside(2.*r - bb)
+            
+    
+    # position of the centers of the balls and the red dots on the ball. Needed for the animation
+    Dmc_loc = []
+    punkt_loc = []
+    for i in range(n):
+        Dmc_loc.append([me.dot(Dmc[i].pos_from(P[0]), uv) for uv in N])  
+        punkt_loc.append([me.dot(punkt[i].pos_from(P[0]), uv) for uv in N])
+    
+    # Lambdification
+    qL = q1 + u1
+    pL = [m, m1, g, r, l, iXX, iYY, iZZ, reibung, k]
+    
+    MM_lam = sm.lambdify(qL + pL, MM, cse=cse1)
+    force_lam = sm.lambdify(qL + pL, force, cse=cse1)
+    
+    pot_lam = sm.lambdify(qL + pL, pot_energie, cse=cse1)
+    kin_lam = sm.lambdify(qL + pL, kin_energie, cse=cse1)
+    spring_lam = sm.lambdify(qL + pL, spring_energie, cse=cse1)
+    
+    Dmc_loc_lam = sm.lambdify(qL + pL, Dmc_loc, cse=cse1)
+    punkt_loc_lam = sm.lambdify(qL + pL, punkt_loc, cse=cse1)
+    
+    print('it took {:.3f} sec to set up Kanes equations'.format(time.time() - start))
```

Comment:
This is one big code cell. If you want to display any intermediate outputs in this code cell, break it up into more code cells with more `.. jupyter-execute::` lines.